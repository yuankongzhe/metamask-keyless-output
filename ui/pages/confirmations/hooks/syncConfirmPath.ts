import { useEffect } from 'react';
import { useHistory, useParams } from 'react-router-dom';
import { useSelector } from 'react-redux';

import {
  CONFIRM_TRANSACTION_ROUTE,
  SIGNATURE_REQUEST_PATH,
} from '../../../helpers/constants/routes';
import { currentConfirmationSelector } from '../../../selectors/confirm';

const syncConfirmPath = () => {
  const history = useHistory();
  const { id: paramsTransactionId } = useParams<{ id: string }>();

  const currentConfirmation = useSelector(currentConfirmationSelector);

  // Redirect below is done to keep the confirmation routes backward compatible
  useEffect(() => {
    if (!currentConfirmation) {
      return;
    }
    if (paramsTransactionId !== currentConfirmation.id) {
      const isSignature = Boolean(currentConfirmation.msgParams);
      if (isSignature) {
        history.replace(
          `${CONFIRM_TRANSACTION_ROUTE}/${currentConfirmation.id}/${SIGNATURE_REQUEST_PATH}`,
        );
      }
      // todo: logic to replace url for rest of transactions to be added here
    }
  }, [currentConfirmation, paramsTransactionId]);
};

export default syncConfirmPath;
