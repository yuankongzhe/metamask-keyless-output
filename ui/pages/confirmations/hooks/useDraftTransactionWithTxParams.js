import { useSelector } from 'react-redux';
import { getCurrentDraftTransaction } from '../../../ducks/send';
import { getUnapprovedTransactions } from '../../../selectors';

/**
 * Returns an object that resembles the txData.txParams from the Transactions state.
 * While processing gas details for send transaction and edit transaction,
 * the gas data from draftTransaction and unapprovedTx has to be reorganized
 * to mimic the txdata.txParam from a confirmTransaction
 *
 * @returns {object} The transaction data
 */
export const useDraftTransactionWithTxParams = () => {
  const draftTransaction = useSelector(getCurrentDraftTransaction);

  const unapprovedTxs = useSelector(getUnapprovedTransactions);

  let transactionData = {};

  if (Object.keys(draftTransaction).length !== 0) {
    const editingTransaction = unapprovedTxs[draftTransaction.id];
    transactionData = {
      txParams: {
        gasPrice: draftTransaction.gas?.gasPrice,
        gas: editingTransaction?.userEditedGasLimit
          ? editingTransaction?.txParams?.gas
          : draftTransaction.gas?.gasLimit,
        maxFeePerGas: editingTransaction?.txParams?.maxFeePerGas
          ? editingTransaction?.txParams?.maxFeePerGas
          : draftTransaction.gas?.maxFeePerGas,
        maxPriorityFeePerGas: editingTransaction?.txParams?.maxPriorityFeePerGas
          ? editingTransaction?.txParams?.maxPriorityFeePerGas
          : draftTransaction.gas?.maxPriorityFeePerGas,
        value: draftTransaction.amount?.value,
        type: draftTransaction.transactionType,
      },
      userFeeLevel: editingTransaction?.userFeeLevel,
    };
  }

  return transactionData;
};
