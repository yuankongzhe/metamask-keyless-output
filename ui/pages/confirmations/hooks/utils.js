import {
  GasEstimateTypes,
  CUSTOM_GAS_ESTIMATE,
} from '../../../../shared/constants/gas';

export function getGasFeeEstimate(
  field,
  gasFeeEstimates,
  gasEstimateType,
  estimateToUse,
  fallback = '0',
) {
  if (gasEstimateType === GasEstimateTypes.feeMarket) {
    return gasFeeEstimates?.[estimateToUse]?.[field] ?? String(fallback);
  }
  return String(fallback);
}

export const feeParamsAreCustom = (transaction) =>
  !transaction?.userFeeLevel ||
  transaction?.userFeeLevel === CUSTOM_GAS_ESTIMATE;
