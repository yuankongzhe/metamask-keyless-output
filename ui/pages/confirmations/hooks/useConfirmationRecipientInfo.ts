import { useSelector } from 'react-redux';

import {
  accountsWithSendEtherInfoSelector,
  currentConfirmationSelector,
} from '../../../selectors';
import { getAccountByAddress } from '../../../helpers/utils/util';

function useConfirmationRecipientInfo() {
  const currentConfirmation = useSelector(currentConfirmationSelector);
  const allAccounts = useSelector(accountsWithSendEtherInfoSelector);

  let recipientAddress = '';
  let recipientName = '';

  if (currentConfirmation) {
    const { msgParams } = currentConfirmation;
    // url for all signature requests
    if (msgParams) {
      recipientAddress = msgParams.from;
      const fromAccount = getAccountByAddress(allAccounts, recipientAddress);
      recipientName = fromAccount?.metadata?.name;
    }
    // TODO: as we add support for more transaction code to find recipient address for different
    // transaction types will come here
  }

  return {
    recipientAddress,
    recipientName,
  };
}

export default useConfirmationRecipientInfo;
