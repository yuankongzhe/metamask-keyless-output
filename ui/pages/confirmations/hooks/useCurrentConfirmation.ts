import { useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';
import { useSelector } from 'react-redux';

import { ApprovalRequest } from '@metamask/approval-controller';
import { Json } from '@metamask/utils';

import {
  latestPendingConfirmationSelector,
  pendingConfirmationsSelector,
  unapprovedPersonalMsgsSelector,
} from '../../../selectors';

type Approval = ApprovalRequest<Record<string, Json>>;

const useCurrentConfirmation = () => {
  const { id: paramsTransactionId } = useParams<{ id: string }>();
  const unapprovedPersonalMsgs = useSelector(unapprovedPersonalMsgsSelector);
  const latestPendingConfirmation: Approval = useSelector(
    latestPendingConfirmationSelector,
  );
  const pendingConfirmations: Approval[] = useSelector(
    pendingConfirmationsSelector,
  );
  const [currentConfirmation, setCurrentConfirmation] =
    useState<Record<string, unknown>>();

  useEffect(() => {
    let pendingConfirmation: Approval | undefined;
    if (paramsTransactionId) {
      if (paramsTransactionId === currentConfirmation?.id) {
        return;
      }
      pendingConfirmation = pendingConfirmations.find(
        ({ id: confirmId }) => confirmId === paramsTransactionId,
      );
    } else {
      pendingConfirmation = latestPendingConfirmation;
    }
    if (
      pendingConfirmation &&
      pendingConfirmation.id !== currentConfirmation?.id
    ) {
      // currently re-design is enabled only for personal signatures
      // condition below can be changed as we enable it for other transactions also
      const unapprovedMsg = unapprovedPersonalMsgs[pendingConfirmation.id];
      if (!unapprovedMsg) {
        return;
      }
      const { siwe } = unapprovedMsg.msgParams;

      if (!siwe?.isSIWEMessage) {
        setCurrentConfirmation(unapprovedMsg);
      }
    }
  }, [latestPendingConfirmation, paramsTransactionId, unapprovedPersonalMsgs]);

  return { currentConfirmation };
};

export default useCurrentConfirmation;
