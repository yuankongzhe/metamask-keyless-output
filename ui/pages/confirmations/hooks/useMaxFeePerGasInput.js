import { useEffect, useState } from 'react';
import { useSelector } from 'react-redux';

import { checkNetworkAndAccountSupports1559 } from '../../../selectors';
import { isLegacyTransaction } from '../../../helpers/utils/transactions.util';

import { hexWEIToDecGWEI } from '../../../../shared/modules/conversion.utils';
import { feeParamsAreCustom, getGasFeeEstimate } from './utils';

const getMaxFeePerGasFromTransaction = (transaction, gasFeeEstimates) => {
  if (gasFeeEstimates?.[transaction?.userFeeLevel]) {
    return gasFeeEstimates[transaction.userFeeLevel].suggestedMaxFeePerGas;
  }
  const { maxFeePerGas, gasPrice } = transaction?.txParams || {};
  return Number(hexWEIToDecGWEI(maxFeePerGas || gasPrice));
};

/**
 * @typedef {object} MaxFeePerGasInputReturnType
 * @property {(DecGweiString) => void} setMaxFeePerGas - state setter method to update the maxFeePerGas.
 * @property {string} [maxFeePerGas] - getter method for maxFeePerGas.
 */

/**
 * @param options
 * @param options.estimateToUse
 * @param options.gasEstimateType
 * @param options.gasFeeEstimates
 * @param options.transaction
 * @returns {MaxFeePerGasInputReturnType}
 */
export function useMaxFeePerGasInput({
  estimateToUse,
  gasEstimateType,
  gasFeeEstimates,
  transaction,
}) {
  const supportsEIP1559 =
    useSelector(checkNetworkAndAccountSupports1559) &&
    !isLegacyTransaction(transaction?.txParams);

  const initialMaxFeePerGas = supportsEIP1559
    ? getMaxFeePerGasFromTransaction(transaction, gasFeeEstimates)
    : 0;

  // This hook keeps track of a few pieces of transitional state. It is
  // transitional because it is only used to modify a transaction in the
  // metamask (background) state tree.
  const [maxFeePerGas, setMaxFeePerGas] = useState(() => {
    if (initialMaxFeePerGas && feeParamsAreCustom(transaction)) {
      return initialMaxFeePerGas;
    }
    return null;
  });

  useEffect(() => {
    if (supportsEIP1559 && initialMaxFeePerGas) {
      setMaxFeePerGas(initialMaxFeePerGas);
    }
  }, [initialMaxFeePerGas, setMaxFeePerGas, supportsEIP1559]);

  // We specify whether to use the estimate value by checking if the state
  // value has been set. The state value is only set by user input and is wiped
  // when the user selects an estimate. Default here is '0' to avoid bignumber
  // errors in later calculations for nullish values.
  const maxFeePerGasToUse =
    maxFeePerGas ??
    getGasFeeEstimate(
      'suggestedMaxFeePerGas',
      gasFeeEstimates,
      gasEstimateType,
      estimateToUse,
      initialMaxFeePerGas || 0,
    );

  return {
    maxFeePerGas: maxFeePerGasToUse,
    setMaxFeePerGas,
  };
}
