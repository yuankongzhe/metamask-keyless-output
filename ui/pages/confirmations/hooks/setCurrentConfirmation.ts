import { useEffect } from 'react';
import { useDispatch } from 'react-redux';

import { updateCurrentConfirmation } from '../../../ducks/confirm/confirm';
import useCurrentConfirmation from './useCurrentConfirmation';

/*
 * This hook is called from <Confirm /> component to set current transaction.
 * This hook should be invoked only when we are setting or resetting the
 * current confirmation displayed to the user.
 */
const setCurrentConfirmation = () => {
  const dispatch = useDispatch();
  const { currentConfirmation } = useCurrentConfirmation();

  useEffect(() => {
    if (currentConfirmation) {
      dispatch(updateCurrentConfirmation(currentConfirmation));
    }
  }, [currentConfirmation]);

  useEffect(() => {
    return () => {
      dispatch(updateCurrentConfirmation(undefined));
    };
  }, []);
};

export default setCurrentConfirmation;
