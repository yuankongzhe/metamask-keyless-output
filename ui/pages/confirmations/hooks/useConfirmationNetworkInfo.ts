import { useSelector } from 'react-redux';

import {
  NETWORK_TO_NAME_MAP,
  NETWORK_TYPES,
} from '../../../../shared/constants/network';

import {
  currentConfirmationSelector,
  getAllNetworks,
} from '../../../selectors';
import { getProviderConfig } from '../../../ducks/metamask/metamask';

import { useI18nContext } from '../../../hooks/useI18nContext';

type KeyOfNetworkName = keyof typeof NETWORK_TO_NAME_MAP;

function useConfirmationNetworkInfo() {
  const currentConfirmation = useSelector(currentConfirmationSelector);
  const allNetworks = useSelector(getAllNetworks);
  const providerConfig = useSelector(getProviderConfig);
  const t = useI18nContext();

  let networkDisplayName = '';
  let confirmationNetwork;

  if (currentConfirmation) {
    // use the current confirmation chainId, else use the current network chainId
    const currentChainId =
      currentConfirmation?.chainId ?? providerConfig.chainId;
    confirmationNetwork = allNetworks.find(
      ({ chainId }) => chainId === currentChainId,
    );
    if (confirmationNetwork) {
      const { nickname, providerType, type } = confirmationNetwork;
      if (providerType === NETWORK_TYPES.RPC || type === NETWORK_TYPES.RPC) {
        networkDisplayName = nickname ?? t('privateNetwork');
      } else {
        networkDisplayName =
          NETWORK_TO_NAME_MAP[confirmationNetwork?.chainId as KeyOfNetworkName];
      }
    }
  }

  return {
    networkImageUrl: confirmationNetwork?.rpcPrefs?.imageUrl ?? '',
    networkDisplayName,
  };
}

export default useConfirmationNetworkInfo;
