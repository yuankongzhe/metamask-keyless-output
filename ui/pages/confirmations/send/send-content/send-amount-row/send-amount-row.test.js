import React from 'react';
import { fireEvent } from '@testing-library/react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import mockSendState from '../../../../../../test/data/mock-send-state.json';
import { renderWithProvider } from '../../../../../../test/lib/render-helpers';
import {
  AssetType,
  TokenStandard,
} from '../../../../../../shared/constants/transaction';
import { useIsOriginalNativeTokenSymbol } from '../../../../../hooks/useIsOriginalNativeTokenSymbol';
import { decimalToHex } from '../../../../../../shared/modules/conversion.utils';
import SendAmountRow from '.';

const mockUpdateSendAmount = jest.fn();

jest.mock('../../../../../ducks/send', () => ({
  ...jest.requireActual('../../../../../ducks/send'),
  updateSendAmount: () => mockUpdateSendAmount,
}));

jest.mock('../../../../../hooks/useIsOriginalNativeTokenSymbol', () => {
  return {
    useIsOriginalNativeTokenSymbol: jest.fn(),
  };
});

jest.mock('../../../../../../shared/modules/conversion.utils', () => ({
  ...jest.requireActual('../../../../../../shared/modules/conversion.utils'),
  decimalToHex: jest
    .fn()
    .mockImplementation((decimal) => `mockedHex-${decimal}`),
}));

describe('SendAmountRow Component', () => {
  useIsOriginalNativeTokenSymbol.mockReturnValue(true);
  describe('render', () => {
    describe('Native Asset Type', () => {
      const mockStore = configureMockStore([thunk])(mockSendState);

      it('should match snapshot for native asset type', () => {
        const { container } = renderWithProvider(<SendAmountRow />, mockStore);

        expect(container).toMatchSnapshot();
      });
    });

    describe('Token Asset Type', () => {
      const tokenState = {
        ...mockSendState,
        send: {
          currentTransactionUUID: '1-tx',
          draftTransactions: {
            '1-tx': {
              asset: {
                balance: '0x1158e460913d00000', // 20000000000000000000
                details: {
                  address: '0x617b3f8050a0BD94b6b1da02B4384eE5B4DF13F4',
                  symbol: 'META',
                  balance: '1000000000000000000',
                  decimals: 18,
                  string: '1',
                  balanceError: null,
                  isERC721: false,
                  image: '',
                  standard: 'ERC20',
                },
                error: null,
                type: AssetType.token,
              },
            },
          },
        },
      };
      const mockStore = configureMockStore([thunk])(tokenState);

      it('should match snapshot for token asset type', () => {
        const { container } = renderWithProvider(<SendAmountRow />, mockStore);

        expect(container).toMatchSnapshot();
      });
    });

    describe('NFT Asset Type', () => {
      it('should match snapshot for token NFT type', () => {
        const nftState = {
          ...mockSendState,
          send: {
            currentTransactionUUID: '1-tx',
            draftTransactions: {
              '1-tx': {
                asset: {
                  balance: '',
                  details: {
                    standard: TokenStandard.ERC721,
                  },
                  error: null,
                  type: AssetType.NFT,
                },
              },
            },
          },
        };

        const mockStore = configureMockStore([thunk])(nftState);

        const { container } = renderWithProvider(<SendAmountRow />, mockStore);

        expect(container).toMatchSnapshot();
      });

      it('should render amount field and balance for erc1155', () => {
        const nftState = {
          ...mockSendState,
          send: {
            currentTransactionUUID: '1-tx',
            draftTransactions: {
              '1-tx': {
                asset: {
                  balance: '',
                  details: {
                    balance: '2',
                    standard: TokenStandard.ERC1155,
                  },
                  error: null,
                  type: AssetType.NFT,
                },
              },
            },
          },
        };

        const mockStore = configureMockStore([thunk])(nftState);

        const { getByText } = renderWithProvider(<SendAmountRow />, mockStore);

        expect(getByText('Amount:')).toBeInTheDocument();
        expect(getByText('Balance: 2 tokens')).toBeInTheDocument();
      });
    });
  });

  describe('updateAmount', () => {
    const mockStore = configureMockStore([thunk])(mockSendState);

    it('should call updateSendAmount', () => {
      const { queryByTestId } = renderWithProvider(
        <SendAmountRow />,
        mockStore,
      );
      queryByTestId('currency-input');

      fireEvent.change(queryByTestId('currency-input'), {
        target: { value: 0.5 },
      });

      expect(mockUpdateSendAmount).toHaveBeenCalled();
    });
  });

  describe('SendAmountRow', () => {
    it('calls decimalToHex on handleChange for erc1155 token', () => {
      const erc1155State = {
        ...mockSendState,
        send: {
          currentTransactionUUID: '1-tx',
          draftTransactions: {
            '1-tx': {
              asset: {
                balance: '',
                details: {
                  standard: TokenStandard.ERC1155,
                },
                error: null,
                type: AssetType.NFT,
              },
            },
          },
        },
      };
      const mockStoreErc1155 = configureMockStore([thunk])(erc1155State);
      const { queryByTestId } = renderWithProvider(
        <SendAmountRow />,
        mockStoreErc1155,
      );

      const input = queryByTestId('token-input');
      fireEvent.change(input, { target: { value: '456' } });

      expect(decimalToHex).toHaveBeenCalledWith('456');
      expect(mockUpdateSendAmount).toHaveBeenCalled();
    });
  });
});
