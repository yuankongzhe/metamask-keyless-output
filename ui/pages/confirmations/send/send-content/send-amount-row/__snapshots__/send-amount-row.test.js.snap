// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendAmountRow Component render NFT Asset Type should match snapshot for token NFT type 1`] = `<div />`;

exports[`SendAmountRow Component render Native Asset Type should match snapshot for native asset type 1`] = `
<div>
  <div
    class="send-v2__form-row"
  >
    <div
      class="send-v2__form-label"
    >
      Amount:
      <button
        class="send-v2__amount-max"
      >
        <input
          readonly=""
          type="checkbox"
        />
        <div
          class="send-v2__amount-max__button"
        >
          Max
        </div>
      </button>
    </div>
    <div
      class="send-v2__form-field-container"
    >
      <div
        class="send-v2__form-field"
      >
        <div
          class="unit-input"
        >
          <div
            class="unit-input__inputs"
          >
            <div
              class="unit-input__input-container"
            >
              <input
                class="unit-input__input"
                data-testid="currency-input"
                dir="ltr"
                placeholder="0"
                style="width: 1.5ch;"
                type="number"
                value="1"
              />
              <div
                class="unit-input__suffix"
              >
                ETH
              </div>
            </div>
            <div
              class="currency-input__conversion-component"
            >
              No conversion rate available
            </div>
          </div>
          <button
            class="currency-input__swap-component"
            data-testid="currency-swap"
          >
            <i
              class="fa fa-retweet fa-lg"
            />
          </button>
        </div>
      </div>
      <div />
    </div>
  </div>
</div>
`;

exports[`SendAmountRow Component render Token Asset Type should match snapshot for token asset type 1`] = `
<div>
  <div
    class="send-v2__form-row"
  >
    <div
      class="send-v2__form-label"
    >
      Amount:
      <button
        class="send-v2__amount-max"
      >
        <input
          readonly=""
          type="checkbox"
        />
        <div
          class="send-v2__amount-max__button"
        >
          Max
        </div>
      </button>
    </div>
    <div
      class="send-v2__form-field-container"
    >
      <div
        class="send-v2__form-field"
      >
        <div
          class="unit-input"
        >
          <div
            class="unit-input__inputs"
          >
            <div
              class="unit-input__input-container"
            >
              <input
                class="unit-input__input"
                dir="ltr"
                placeholder="0"
                style="width: 1.5ch;"
                type="number"
                value="0"
              />
              <div
                class="unit-input__suffix"
              >
                META
              </div>
            </div>
            <div
              class="currency-input__conversion-component"
            >
              No conversion rate available
            </div>
          </div>
        </div>
      </div>
      <div />
    </div>
  </div>
</div>
`;
